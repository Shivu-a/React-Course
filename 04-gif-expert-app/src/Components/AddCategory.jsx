import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState("");

  const onChangedInput = (e) => {
    setInputValue(e.target.value);
  };

  const onSubmit = (e) => {
    // Prevenimos el default
    e.preventDefault();

    // Creamos una variable con el input
    const newInputValue = inputValue.trim();

    // Chequeamos el input
    if (newInputValue.length <= 1) return;

    // Creamos la nueva categoria y reseteamos el value
    onNewCategory(newInputValue);
    setInputValue("");
  };

  return (
    <form onSubmit={onSubmit} aria-label="form">
      <input
        type="text"
        placeholder="Ingrese su busqueda"
        value={inputValue}
        onChange={onChangedInput}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,
};
