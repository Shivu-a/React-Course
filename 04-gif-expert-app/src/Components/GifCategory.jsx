import React from "react";

export const GifCategory = ({ url, id, title }) => {
  return (
    <div className="card">
      <p>{title}</p>
      <img src={url} />
    </div>
  );
};
