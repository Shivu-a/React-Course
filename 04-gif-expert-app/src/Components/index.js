export * from "./AddCategory";
export * from "./GifCategory";
export * from "./GifGrid";
