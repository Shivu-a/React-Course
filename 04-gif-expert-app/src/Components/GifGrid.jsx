import React from "react";

import { GifCategory } from "./GifCategory";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({ category }) => {
  const { images, isLoading } = useFetchGifs(category);

  return (
    <div className="gifCategory">
      <h3>{category}</h3>
      {isLoading && <div className="loading"></div>}
      <div className="card-grid">
        {images.map((image) => (
          <GifCategory key={image.id} {...image} />
        ))}
      </div>
    </div>
  );
};
