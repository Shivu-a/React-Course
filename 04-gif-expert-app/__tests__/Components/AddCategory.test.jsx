import { fireEvent, render, screen } from "@testing-library/react";

import { AddCategory } from "../../src/Components/AddCategory";

describe("Pruebas en AddCategory", () => {
  test("AddCategory deberia hacer match con el snapshot", () => {
    const { container } = render(<AddCategory onNewCategory={() => {}} />);

    expect(container).toMatchSnapshot();
  });

  test("Deberia de cambiar el valor de la caja de texto", () => {
    render(<AddCategory onNewCategory={() => {}} />);

    const input = screen.getByRole("textbox");

    fireEvent.input(input, { target: { value: "Rimuru" } });
    expect(input.value).toBe("Rimuru");
  });

  test("El valor de la caja de texto deberia de volver a ser null despues de hacer submit", () => {
    render(<AddCategory onNewCategory={() => {}} />);

    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    fireEvent.input(input, { target: { value: "Rimuru" } });
    expect(input.value).toBe("Rimuru");

    fireEvent.submit(form);
    expect(input.value).toBeFalsy();
  });

  test("onChangedInput deberia para ser llamado", () => {
    const inputValue = "Rimuru";
    const onChangedInput = jest.fn();

    render(<AddCategory onNewCategory={onChangedInput} />);

    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    fireEvent.input(input, { target: { value: inputValue } });
    expect(input.value).toBe(inputValue);

    fireEvent.submit(form);

    expect(onChangedInput).toHaveBeenCalledTimes(1);
    expect(onChangedInput).toHaveBeenCalledWith(inputValue);
  });

  test("onChangedInput deberia NO ser llamado ya que el input esta vacio", () => {
    const onChangedInput = jest.fn();
    render(<AddCategory onNewCategory={onChangedInput} />);

    const form = screen.getByRole("form");
    fireEvent.submit(form);

    expect(onChangedInput).toHaveBeenCalledTimes(0);
  });
});
