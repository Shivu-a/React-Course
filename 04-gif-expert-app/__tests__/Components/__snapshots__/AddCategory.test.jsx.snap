// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AddCategory AddCategory deberia hacer match con el snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Ingrese su busqueda"
      type="text"
      value=""
    />
  </form>
</div>
`;
